rain_api_core/logging.py

```diff
@@ -0,0 +1,234 @@
+
+    logger = logging.getLogger()
+
+    for h in logger.handlers:
+        logger.removeHandler(h)
+
+    handler = logging.StreamHandler(sys.stdout)
+    handler.setFormatter(formatter)
+    handler.addFilter(custom_log_filter)
+
+    logger.addHandler(handler)
+    logger.setLevel(loglevel)
+
+    if os.getenv("QUIETBOTO", 'TRUE').upper() == 'TRUE':
+        # BOTO, be quiet plz
+        logging.getLogger('boto3').setLevel(logging.ERROR)
+        logging.getLogger('botocore').setLevel(logging.ERROR)
+        logging.getLogger('nose').setLevel(logging.ERROR)
+        logging.getLogger('elasticsearch').setLevel(logging.ERROR)
+        logging.getLogger('s3transfer').setLevel(logging.ERROR)
+        logging.getLogger('urllib3').setLevel(logging.ERROR)
+        logging.getLogger('connectionpool').setLevel(logging.ERROR)
+    return logger
+
+
+class PercentPlaceholder():
+    """A placeholder in a log format object
+
+    The placeholder can be formatted with the % operator.
+
+    >>> p = PercentPlaceholder("message")
+    >>> assert p % {"message": "hello"} == "hello"
+    """
+
+    __slots__ = ("name", )
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def __mod__(self, args):
+        if isinstance(args, dict):
+            return args[self.name]
+
+
+class JSONPercentStyle():
+    default_format = {"message": "%(message)s"}
+    asctime_search = '%(asctime)'
+    placeholder_pattern = re.compile(r"^%\((\w+)\)s$")
+    validation_pattern = re.compile(r'%\(\w+\)[#0+ -]*(\*|\d+)?(\.(\*|\d+))?[diouxefgcrsa%]', re.I)
+
+    def __init__(self, fmt: dict):
+        self._fmt = self._convert_placeholders(fmt or self.default_format)
+        # Cached so we don't recompute it every time a record is emitted
+        self._uses_time = self._usesTime()
+
+    def _convert_placeholders(self, obj):
+        """Convert '%(name)s' values into PercentPlaceholder objects"""
+        def func(obj):
+            if isinstance(obj, str):
+                m = self.placeholder_pattern.match(obj)
+                if m:
+                    return PercentPlaceholder(m.group(1))
+            return obj
+
+        return _map_json_object(func, obj)
+
+    def _usesTime(self):
+        for val in _walk_json_values(self._fmt):
+            if isinstance(val, str) and self.asctime_search in val:
+                return True
+            if isinstance(val, PercentPlaceholder) and val.name == "asctime":
+                return True
+        return False
+
+    def usesTime(self):
+        return self._uses_time
+
+    def _format(self, record):
+        return _fmt_json_object(self._fmt, record.__dict__)
+
+    def format(self, record):
+        try:
+            return self._format(record)
+        except KeyError as e:
+            raise ValueError('Formatting field not found in record: %s' % e)
+
+
+class JSONFormatter(logging.Formatter):
+    def __init__(self, fmt=None, datefmt: str = None):
+        # Changing the type from the base class
```

Comment:
Why no call to super?

Reply:
Nevermind, it pretty much duplicates the super's init

Reply:
The base class doesn't let you inject your own 'style' objects so I wanted to skip all of the code that does the validation for that. Now that I think of it though, overriding a 'private' member is not good practice. I may just be better off not extending `logging.Formatter` at all.

Even though the logging module is written in such a way that you can customize pretty much everything, one thing that's deeply ingrained is the assumption that formatting always happens on (and returns) strings, even intermediate formatting operations.